src/pretix/static/pretixbase/js/addressform.js

```diff
@@ -0,0 +1,63 @@
+
+    $("select[data-country-information-url]").each(function () {
+        let counter = 0;
+        const dependency = $(this),
+            url = this.getAttribute('data-country-information-url'),
+            form = $(this).closest(".panel-body, form, .profile-scope"),
+            isRequired = dependency.closest(".form-group").is(".required"),
+            dependents =  {
+                'city': form.find("input[name$=city]"),
+                'zipcode': form.find("input[name$=zipcode]"),
+                'street': form.find("textarea[name$=street]"),
+                'state': form.find("select[name$=state]"),
+                'vat_id': form.find("input[name$=vat_id]"),
+            },
+            update = function (ev) {
+                counter++;
+                const curCounter = counter;
+                for (var k in dependents) dependents[k].prop("disabled", true);
+                dependency.closest(".form-group").find("label").prepend("<span class='fa fa-cog fa-spin'></span> ");
+                $.getJSON(url + '?country=' + dependency.val(), function (data) {
+                    if (counter > curCounter) {
+                        return;  // Lost race
+                    }
+                    var selected_value = dependents.state.prop("data-selected-value");
+                    if (selected_value) dependents.state.prop("data-selected-value", "");
+                    dependents.state.find("option").filter(function (t) {return !!$(this).attr("value")}).remove();
+                    if (data.data.length > 0) {
+                        $.each(data.data, function (k, s) {
+                            var o = $("<option>").attr("value", s.code).text(s.name);
+                            if (s.code == selected_value || (selected_value && selected_value.indexOf && selected_value.indexOf(s.code) > -1)) {
+                                o.prop("selected", true);
+                            }
+                            dependents.state.append(o);
+                        });
+                    }
+                    for(var k in dependents) {
+                        const options = data[k], dependent = dependents[k];
+                        if ('visible' in options) {
+                            if (options.visible) {
+                                dependent.closest(".form-group").show().toggleClass('required', isRequired);
+                                dependent.prop('required', isRequired);
+                            } else {
+                                dependent.closest(".form-group").hide();
+                                dependent.prop("required", false);
+                            }
+                        }
+                        if ('required' in options) {
+                            dependent.closest(".form-group").toggleClass('required', options.required && isRequired);
+                            dependent.prop('required', options.required && isRequired);
+                        }
```

Comment:
I’d remove the `if ('visible' in options)` as if one country happens to hide the street, then it will not show up again as the js_helper does not always provide the key `visible` in its reponse. To make it more robust, I would add a default-visibility of true like so – also, I made the code more compact:

```suggestion
                        const options = data[k],
                            dependent = dependents[k],
                            visible = 'visible' in options ? options.visible : true,
                            required = 'required' in options && options.required && isRequired && visible;

                        dependent.closest(".form-group").toggle(visible).toggleClass('required', required);
                        dependent.prop("required", required);
```

Reply:
> as the js_helper does not always provide the key visible in its reponse

How so? The js_helper always returns the exact same set of keys, just different values. So all fields that can get hidden also will be made visible again, if required. 

But I like your more compact logic better, I'll just have to double check whether it really behaves the same (I think so, but need to try it - or did you already test it?)

Reply:
You are right, currently there cannot be any setup where e.g. street would get a key `visible` only in the case it is hidden. But I guess to make it fail-safe, it might be a good idea to default to something, so if something else hid the element, it gets unhidden.

I did test it locally, but it adds one more thing: an input, that is hidden, never can be required – that is why the `&& visible` is at the end of the reuqired-assignment. I think that is what we want, but wanted to point that out.